resources/assets/lib/beatmap-discussions/new-discussion.tsx

```diff
@@ -0,0 +1,517 @@
+import BigButton from 'components/big-button';
+import StringWithComponent from 'components/string-with-component';
+import TimeWithTooltip from 'components/time-with-tooltip';
+import UserAvatar from 'components/user-avatar';
+import BeatmapExtendedJson from 'interfaces/beatmap-extended-json';
+import BeatmapsetDiscussionJson from 'interfaces/beatmapset-discussion-json';
+import { BeatmapsetDiscussionPostStoreResponseJson } from 'interfaces/beatmapset-discussion-post-responses';
+import BeatmapsetExtendedJson from 'interfaces/beatmapset-extended-json';
+import { BeatmapsetWithDiscussionsJson } from 'interfaces/beatmapset-json';
+import CurrentUserJson from 'interfaces/current-user-json';
+import GameMode from 'interfaces/game-mode';
+import { route } from 'laroute';
+import { action, computed, makeObservable, observable, runInAction } from 'mobx';
+import { observer } from 'mobx-react';
+import core from 'osu-core-singleton';
+import * as React from 'react';
+import TextareaAutosize from 'react-autosize-textarea';
+import { onError } from 'utils/ajax';
+import { canModeratePosts, validMessageLength } from 'utils/beatmapset-discussion-helper';
+import { nominationsCount } from 'utils/beatmapset-helper';
+import { classWithModifiers } from 'utils/css';
+import { InputEventType, makeTextAreaHandler } from 'utils/input-handler';
+import { hideLoadingOverlay, showLoadingOverlay } from 'utils/loading-overlay';
+import { linkHtml } from 'utils/url';
+import MessageLengthCounter from './message-length-counter';
+
+const bn = 'beatmap-discussion-new';
+
+type Discussion = BeatmapsetDiscussionJson & Required<Pick<BeatmapsetDiscussionJson, 'current_user_attributes'>>;
+type Mode = 'events' | 'general' | 'generalAll' | 'timeline' | 'reviews';
+
+// TODO: move to store/context
+interface CurrentDiscussions {
+  byFilter: {
+    deleted: DiscussionByFilter;
+    hype: DiscussionByFilter;
+    mapperNotes: DiscussionByFilter;
+    mine: DiscussionByFilter;
+    pending: DiscussionByFilter;
+    praises: DiscussionByFilter;
+    resolved: DiscussionByFilter;
+    total: DiscussionByFilter;
+  };
+  countsByBeatmap: Record<number, number>;
+  countsByPlaymode: Record<GameMode, number>;
+  general: Discussion[];
+  generalAll: Discussion[];
+  reviews: Discussion[];
+  timeline: Discussion[];
+  timelineAllUsers: Discussion[];
+  totalHype: number;
+  unresolvedIssues: number;
+}
+
+interface DiscussionByFilter {
+  general: Record<number, Discussion>;
+  generalAll: Record<number, Discussion>;
+  reviews: Record<number, Discussion>;
+  timeline: Record<number, Discussion>;
+}
+
+interface DiscussionsCache {
+  beatmap: BeatmapExtendedJson;
+  discussions: Discussion[];
+  timestamp: number | null;
+}
+
+interface Props {
+  autoFocus: boolean;
+  beatmapset: BeatmapsetExtendedJson & BeatmapsetWithDiscussionsJson;
+  currentBeatmap: BeatmapExtendedJson;
+  currentDiscussions: CurrentDiscussions;
+  currentUser: CurrentUserJson;
```

Comment:
there's a weird mixture of `this.props.currentUser` and `core.currentUser` going on here ಠ_ಠ use the latter for everything instead?